rfd/0054-passwordless-macos.md

```diff
@@ -0,0 +1,424 @@
+    resident keys, even if the RPID/Teleport where to request
+    [ResidentKeyRequirement = "discouraged"](
+    https://www.w3.org/TR/webauthn-2/#enum-residentKeyRequirement).
+
+1. If registration is successful, `tsh` replaces any existing keys for the
+   RPID+user pair with the newly-created key. This simplifies the authentication
+   ceremony and allows re-registration as a fallback mechanism.
+
+A few parameters specified in the code example deserve note:
+
+* [kSecAttrAccessibleWhenUnlockedThisDeviceOnly](
+  https://developer.apple.com/documentation/security/ksecattraccessiblewhenunlockedthisdeviceonly?language=objc)
+  requires the device to be unlocked and the user to have a password set. It is
+  the more restrictive of the possible "Accessibility Values".
+
+* [kSecAccessControlBiometryAny](
+  https://developer.apple.com/documentation/security/secaccesscontrolcreateflags/ksecaccesscontrolbiometryany)
+  requires a biometric check (Touch ID on macOS). It is more restrictive than
+  *kSecAccessControlUserPresence* (which allows passwords), but less restrictive
+  than *kSecAccessControlBiometryCurrentSet* (doesn't work with newly enrolled
+  fingerprints). *kSecAccessControlBiometryAny* seems to be the sweet spot of
+  security and usability.
+
+In case of a registration failure, `tsh` must do its best to delete the
+created-but-not-registered credential. If all fails, it is possible to use the
+hidden [`tsh` support commands](#tsh-support-commands) for a manual cleanup.
+
+### Authentication
+
+Authentication offers a plethora of options, depending on both server settings
+(otp, webauthn, passwordless) and client state (FIDO2 keys present, Touch ID
+keys registered). In order to decide which flow to follow, `tsh` must first
+assess what is possible, preferably without asking for unnecessary user
+interaction.
+
+Unlike FIDO2 keys, it is possible for `tsh` to discover if Touch ID keys are
+registered in the Enclave _without_ user interaction. Because all Touch ID keys
+are functionally resident keys, as long as the server supports passwordless,
+then `tsh` is free to use it.
+
+If Touch ID passwordless authentication is possible, then it is the preferred
+authentication mode and is used by default.
+
+If passwordless is not possible, but Touch ID is present, then Touch ID MFA is
+also preferred and used by default.
+
+To allow users agency over the eager behaviors of Touch ID, `tsh` is augmented
+with the following flags:
+
+* `tsh --pwdless_mode={auto,on,off}` - activate/deactivate passwordless
+  authentication
+
+    `auto` is the default behavior described above (passwordless is preferred
+    if we are certain it may be used)
+
+    `on` enables passwordless logins (`tsh login --pwdless`, described in the
+    [Passwordless FIDO2 RFD][passwordless fido2], is a shortcut to it)
+
+    `off` disables passwordless logins
+
+* `tsh --mfa_mode={auto,platform,cross-platform}` - choose whether to use
```

Comment:
Can we make auto choice here? For MacOS make this `TouchID` for Linux, make this `FIDO2` and remove the flag?

Reply:
The reasoning for the `--mfa_mode` flag is similar to https://github.com/gravitational/teleport/pull/9296#discussion_r797730723 - `auto` locks the user on Touch ID, but they may want to use a FIDO2 key for some reason, so this exists to allow that. We could make the flag hidden, but as in the other thread, I do find it useful to have.

That said, I would expect most users to rely on the default/auto behavior and never touch this.